File: week6/c++/week6_6.cpp
#include <iostream>
#include <vector>
#include <math.h>

using namespace std;

int main(int argc, char **argv)
{
    int test_case;
    int T;
    cin >> T;

    for (test_case = 1; test_case <= T; ++test_case)
    {
        int k;
        cin >> k;
        vector<vector<int>> num(k + 1, vector<int>());
        for (int i = 0; i < pow(2, k); i++)
        {
            int n;
            cin >> n;
            num[k].push_back(n);
        }

        int sum = 0;
        for (int i = k; i > 0; i--)
        {
            for (int j = 0; j < pow(2, i); j = j + 2)
            {
                int diff, next;
                if (num[i][j] >= num[i][j + 1])
                {
                    diff = num[i][j] - num[i][j + 1];
                    next = num[i][j];
                }
                else
                {
                    diff = num[i][j + 1] - num[i][j];
                    next = num[i][j + 1];
                }
                sum += diff;
                num[i - 1].push_back(next);
            }
        }

        cout << "#" << test_case << " ";
        cout << sum << endl;
    }
}

File: week6/java/week6_6.java
import java.util.Scanner;

public class week6_6 {
    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();

        for(int test_case=1; test_case<=t; test_case++){
            int k = sc.nextInt();

            int[][] arr = new int[k+1][(int)Math.pow(2, k)];
            for(int i=0; i<Math.pow(2, k); i++){
                arr[0][i] = sc.nextInt();
            }

            int answer = 0;
            for(int i=1; i<=k; i++){
                int len = arr[0].length/i;
                int a = 0;
                for(int j=0; j<len; j+=2){
                    if(arr[i-1][j] <= arr[i-1][j+1]){
                        arr[i][a] = arr[i-1][j+1];
                        answer += arr[i-1][j+1] - arr[i-1][j];
                    }else{
                        arr[i][a] = arr[i-1][j];
                        answer += arr[i-1][j] - arr[i-1][j+1];
                    }
                    a++;
                }
            }

            System.out.println("#"+test_case+" "+answer);
        }
    }
}